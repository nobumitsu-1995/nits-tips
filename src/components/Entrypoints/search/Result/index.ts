export { Container as Result } from './Container'
