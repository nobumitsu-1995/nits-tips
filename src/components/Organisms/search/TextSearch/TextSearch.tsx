import React, { useId, useState } from 'react'
import { Heading } from '@Atoms/Heading'
import * as styles from './TextSearch.css'
import { SearchForm } from '../SearchForm'
import {
  SearchTypeSelector,
  SEARCH_TYPE,
  type SearchType,
} from './SearchTypeSelector'

export const TextSearch: React.FC = () => {
  const [searchType, setSearchType] = useState<SearchType>(SEARCH_TYPE.all)
  const [searchWord, setSearchWord] = useState('')
  const a11yId = useId()
  const handleChangeSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchWord(e.target.value)
  }
  return (
    <section>
      <Heading id={a11yId} text="ワード検索" />
      <div className={styles.formContainer}>
        <SearchTypeSelector
          searchType={searchType}
          setSearchType={setSearchType}
        />
        <SearchForm
          id={a11yId}
          searchType={searchType}
          searchWord={searchWord}
          handleChangeSearch={handleChangeSearch}
        />
      </div>
    </section>
  )
}
