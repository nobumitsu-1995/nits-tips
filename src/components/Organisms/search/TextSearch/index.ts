export { TextSearch } from './TextSearch'
export { SEARCH_TYPE, type SearchType } from './SearchTypeSelector'
