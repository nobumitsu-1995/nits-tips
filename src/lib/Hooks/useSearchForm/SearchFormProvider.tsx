import React from 'react'
import { createContext, useContext, type PropsWithChildren } from 'react'
import {
  useSearchForm,
  type UseSearchFormPayloadType,
  type UseSearchFormReturnType,
} from './useSearchForm'

type Common = 'articles' | 'isLoading'
type SearchFormAction = Omit<UseSearchFormReturnType, Common> | null
type SearchFormState = Pick<UseSearchFormReturnType, Common> | null

const SearchFormActionContext = createContext<SearchFormAction>(null)
const SearchFormStateContext = createContext<SearchFormState>(null)
const ActionProvider = SearchFormActionContext.Provider
const StateProvider = SearchFormStateContext.Provider

export const SearchFormProvider = ({
  initialArticles,
  initialTags,
  initialCategory,
  initialSearchType,
  initialSearchWord,
  children,
}: PropsWithChildren<UseSearchFormPayloadType>) => {
  const {
    articles,
    isLoading,
    handleSetCategory,
    handleSetTags,
    handleChangeSortType,
    handleSetSearch,
    handleSetSearchType,
    handleSubmitSearch,
  } = useSearchForm({
    initialArticles,
    initialTags,
    initialCategory,
    initialSearchType,
    initialSearchWord,
  })

  return (
    <ActionProvider
      value={{
        handleSetCategory,
        handleSetTags,
        handleChangeSortType,
        handleSetSearch,
        handleSetSearchType,
        handleSubmitSearch,
      }}
    >
      <StateProvider
        value={{
          articles,
          isLoading,
        }}
      >
        {children}
      </StateProvider>
    </ActionProvider>
  )
}

export const useSearchFormAction = useContext(SearchFormActionContext)
export const useSearchFormState = useContext(SearchFormStateContext)
