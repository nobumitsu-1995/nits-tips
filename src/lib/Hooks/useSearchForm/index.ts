export {
  SearchFormProvider,
  useSearchFormAction,
  useSearchFormState,
} from './SearchFormProvider'
