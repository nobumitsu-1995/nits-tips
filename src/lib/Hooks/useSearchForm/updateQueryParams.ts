import type { SearchType } from '@/components/Organisms/search/TextSearch'
import { generateSearchUrlQuery } from '@/lib/helpers/generateSearchUrlQuery'

type FilterSearch = {
  tags: string[]
  category: string
}

type WordSearch = {
  searchType: SearchType
  searchWord: string
}

type Args = FilterSearch | WordSearch

export const updateQueryParams = (args: Args) => {
  const newQueryParamValue = generateSearchUrlQuery(args)
  const newUrl = `${window.location.pathname}?${newQueryParamValue}`
  window.history.replaceState(null, '', newUrl)
}
