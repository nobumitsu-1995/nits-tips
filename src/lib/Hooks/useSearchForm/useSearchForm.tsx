import { useEffect, useRef, useState } from 'react'
import { useFilterSearch } from '../useFilterSearch'
import { useWordSearch } from '../useWordSearch'
import type {
  UseFilterSearchPayloadType,
  UseFilterSearchReturnType,
} from '../useFilterSearch/useFilterSearch'
import type {
  UseWordSearchPayloadType,
  UseWordSearchReturnType,
} from '../useWordSearch/useWordSearch'
import { convertToArticles } from '@/lib/helpers/convertToArticles'
import { getBlogsData } from '@/lib/API/microCMS/getBlogs'
import type { ArticleCardData } from '@/lib/interfaces/Article'

export type UseSearchFormPayloadType = {
  initialArticles: ArticleCardData[]
} & UseFilterSearchPayloadType &
  UseWordSearchPayloadType

export type UseSearchFormReturnType = {
  articles: ArticleCardData[]
  isLoading: boolean
  handleSubmitSearch: (e: React.FormEvent<HTMLFormElement>) => void
} & Omit<UseFilterSearchReturnType, 'filterParams'> &
  Omit<UseWordSearchReturnType, 'wordParams'>

export const useSearchForm = ({
  initialArticles,
  initialTags,
  initialCategory,
  initialSearchType,
  initialSearchWord,
}: UseSearchFormPayloadType): UseSearchFormReturnType => {
  const isFirstRender = useRef(true)
  const [articles, setArticles] = useState(initialArticles)
  const [isLoading, setIsLoading] = useState(false)

  const {
    filterParams,
    selectedFilters,
    handleSetCategory,
    handleSetTags,
    handleChangeSortType,
  } = useFilterSearch({ initialTags, initialCategory })
  const { wordParams, selectedWord, handleSetSearch, handleSetSearchType } =
    useWordSearch({
      initialSearchType,
      initialSearchWord,
    })

  /** 絞り込み検索の実行処理 */
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false
      return
    }
    setIsLoading(true)
    getBlogsData({
      ...filterParams,
      limit: 100,
    })
      .then((res) => setArticles(convertToArticles(res ?? [])))
      .finally(() => setIsLoading(false))
  }, [filterParams])

  /** ワード検索実行時に発火する関数 */
  const handleSubmitSearch = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    getBlogsData({
      ...wordParams,
      limit: 100,
    })
      .then((res) => setArticles(convertToArticles(res ?? [])))
      .finally(() => setIsLoading(false))
  }

  return {
    articles,
    isLoading,
    selectedFilters,
    selectedWord,
    handleSetCategory,
    handleSetTags,
    handleChangeSortType,
    handleSetSearch,
    handleSetSearchType,
    handleSubmitSearch,
  }
}
