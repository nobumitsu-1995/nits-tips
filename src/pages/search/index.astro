---
import { FilterSearch } from '@/components/Organisms/search/FilterSearch'
import { TextSearch } from '@/components/Organisms/search/TextSearch'
import { Border } from '@/components/UI/Atoms/Border'
import { Card } from '@/components/UI/Atoms/Card'
import BaseLayout from '@layout/BaseLayout.astro'
import { getAllCategories } from '@/lib/API/microCMS/getAllCategories'
import { getAllTags } from '@/lib/API/microCMS/getAllTags'
import * as styles from '@/styles/pages/search.css'

export const prerender = false

const pageTitle = 'nits-tips | 検索画面'
const description = 'nobumitsu-1995による技術tipsの備忘録集。検索画面です。'
const categories = await getAllCategories()
const tags = await getAllTags()
---

<BaseLayout pageTitle={pageTitle} description={description}>
  <main class={styles.container}>
    <div class={styles.cardContainer}>
      <Card padding='48px 28px'>
        <TextSearch client:load />
        <Border />
        <FilterSearch client:load categoriesData={categories} tagsData={tags} />
      </Card>
    </div>
  </main>
</BaseLayout>
