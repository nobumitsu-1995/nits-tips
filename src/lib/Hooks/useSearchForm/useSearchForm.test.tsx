import React from 'react'
import { waitFor } from '@testing-library/react'
import { useSearchForm } from './useSearchForm'
import { createArtilceArray } from '@/testUtils/fixtures/articleData'
import { SEARCH_TYPE } from '@/components/Organisms/search/TextSearch'
import { setup } from '@/testUtils'

const Component = () => {
  const { isLoading, handleSubmitSearch, handleSetTags } = useSearchForm({
    initialArticles: createArtilceArray(),
    initialTags: ['testTag'],
    initialCategory: 'testCategory',
    initialSearchType: SEARCH_TYPE.all,
    initialSearchWord: 'search',
  })

  return (
    <div>
      <p data-testid="loading">{`${isLoading}`}</p>
      <label htmlFor="test">
        testTag2
        <input
          id="test"
          type="checkbox"
          defaultChecked={false}
          onChange={handleSetTags}
          value="testTag2"
        />
      </label>
      <form onSubmit={handleSubmitSearch}>
        <button type="submit">button</button>
      </form>
    </div>
  )
}

const mockedJson = vitest.fn().mockResolvedValue({})
const mockedFetch = vitest.fn()

beforeEach(() => {
  const mockResponse = {
    ok: true,
    status: 200,
    json: mockedJson,
  }
  global.fetch = mockedFetch.mockResolvedValue(mockResponse)
})

afterEach(() => {
  vitest.resetAllMocks()
})

test('filterParamsが更新された時、fetchが走り記事を取得する', async () => {
  expect.assertions(5)
  const { getByRole, getByTestId, user } = setup(<Component />)
  user.click(getByRole('checkbox', { name: 'testTag2' }))

  await waitFor(() => {
    expect(getByTestId('loading').textContent).toBe('true')
  })

  await waitFor(() => {
    expect(getByTestId('loading').textContent).toBe('false')
    expect(mockedFetch).toHaveBeenCalledWith(
      expect.stringContaining(
        'blogs?limit=100&filters=isHidden%5Bnot_equals%5Dtrue%5Band%5Dcategory%5Bequals%5DtestCategory%5Band%5Dtags%5Bcontains%5DtestTag%5Band%5Dtags%5Bcontains%5DtestTag2&orders=-updatedAt',
      ),
      expect.any(Object),
    )
    expect(mockedJson).toHaveBeenCalledWith()
  })
})

test('handleSubmitSearchが呼ばれたときにisLoadingがtrueになり、後でfalseになること', async () => {
  expect.assertions(5)
  const { getByRole, getByTestId, user } = setup(<Component />)
  user.click(getByRole('button', { name: 'button' }))

  await waitFor(() => {
    expect(getByTestId('loading').textContent).toBe('true')
  })

  await waitFor(() => {
    expect(getByTestId('loading').textContent).toBe('false')
    expect(mockedFetch).toHaveBeenCalledWith(
      expect.stringContaining(
        'blogs?limit=100&filters=isHidden%5Bnot_equals%5Dtrue&q=search',
      ),
      expect.any(Object),
    )
    expect(mockedJson).toHaveBeenCalledWith()
  })
})
