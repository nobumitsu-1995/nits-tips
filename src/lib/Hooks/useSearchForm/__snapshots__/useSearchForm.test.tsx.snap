// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filterParamsが更新された時、fetchが走り記事を取得する 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "https://nits-tips.microcms.io/api/v1/blogs?orders=-updatedAt&limit=100&filters=isHidden%5Bnot_equals%5Dtrue%5Band%5Dcategory%5Bequals%5DtestCategory%5Band%5Dtags%5Bcontains%5DtestTag%5Band%5Dtags%5Bcontains%5DtestTag2",
      {
        "headers": Headers {
          Symbol(entries): {
            "x-microcms-api-key": {
              "name": "X-MICROCMS-API-KEY",
              "value": [
                "sample",
              ],
            },
          },
        },
        "method": "GET",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "json": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": {},
            },
          ],
        },
        "ok": true,
        "status": 200,
      },
    },
  ],
}
`;

exports[`handleSubmitSearchが呼ばれたときにisLoadingがtrueになり、後でfalseになること 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "https://nits-tips.microcms.io/api/v1/blogs?orders=-publishedAt&limit=100&filters=isHidden%5Bnot_equals%5Dtrue&q=search",
      {
        "headers": Headers {
          Symbol(entries): {
            "x-microcms-api-key": {
              "name": "X-MICROCMS-API-KEY",
              "value": [
                "sample",
              ],
            },
          },
        },
        "method": "GET",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "json": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "ok": true,
        "status": 200,
      },
    },
  ],
}
`;
